flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -1160,14 +1160,25 @@ public final CloseableRegistry getCancelables() {
 		return recordWriters;
 	}
 
+	@SuppressWarnings("unchecked")
 	private static <OUT> RecordWriter<SerializationDelegate<StreamRecord<OUT>>> createRecordWriter(
 			StreamEdge edge,
 			int outputIndex,
 			Environment environment,
 			String taskName,
 			long bufferTimeout) {
-		@SuppressWarnings("unchecked")
-		StreamPartitioner<OUT> outputPartitioner = (StreamPartitioner<OUT>) edge.getPartitioner();
+
+		StreamPartitioner<OUT> outputPartitioner = null;
+
+		// Clones the partition to avoid multiple stream edges sharing the same stream partitioner,
+		// like the case of https://issues.apache.org/jira/browse/FLINK-14087.
+		try {
+			outputPartitioner = InstantiationUtil.clone(
+				(StreamPartitioner<OUT>) edge.getPartitioner(),
+				Thread.currentThread().getContextClassLoader());
```

Comment:
I am really sorry for yet another comment, but I've spotted it only now.

How about we use the `environment.getUserClassLoader()` instead of `Thread.currentThread().getContextClassLoader()`? I think it is safer that way.

Reply:
Very thanks Dawid! I also think that it would be much better that we get the user classloader in a more deterministic way. I have updated the PR.